proxy/src/main/java/com/velocitypowered/proxy/plugin/util/DirectedAcyclicGraph.java

```diff
@@ -0,0 +1,145 @@
+package com.velocitypowered.proxy.plugin.util;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.*;
+
+public class DirectedAcyclicGraph<T> {
```

Comment:
Maybe you can make use of Guava's graph API instead here?

Reply:
Admittedly, he probably ripped most of this from Voxelwind. Guava 19 (which Voxelwind used) didn't have a graph API.